### ISSUE TITLE
"AnnData.concatenate" discards attributes

### ISSUE THREAD

**Issue by chris-rands** [theislab/anndata#263]
The [code says](https://github.com/theislab/anndata/blob/b0e1a004e64d0e3aadbb9265e0a2fd7ead27b416/anndata/core/anndata.py#L1826) that:
```The :attr:`uns`, :attr:`varm` and :attr:`obsm` attributes are ignored.```

However, it other attributes are also discarded

Practical example:

```
>>> adata
View of AnnData object with n_obs × n_vars = 1525 × 52551 
    var: 'gene_ids', 'feature_types'
>>> adata.raw = adata
>>> adata.raw
<anndata.core.anndata.Raw object at 0x12f967748>
>>> adata_concat = AnnData.concatenate(adata, adata)
>>> adata_concat.raw
```
Toy example:
```
>>> adata.name = 'Chris'
>>> adata_concat = AnnData.concatenate(adata, adata)
>>> adata_concat.name
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'AnnData' object has no attribute 'name'
```

Is this intended or not? I was surprised that attributes were discarded. In my case I accidentally lost the `.raw` data, although of course I can work around it

**Comment by flying-sheep** [comment]
The problem here is that it’s not clear which raw to preserve if there are multiple.

Sure, if they’re equal or only one has a .raw, there’s no problem, but it’s unclear what to do if there’ multiple different .raws.

**Comment by chris-rands** [comment]
Thanks, my expectation might be flawed, but I thought that the raw expression values in `raw.X` could also be concatenated in the same way that `.X` is currently?

For context, I have an `adata` object with `adata.raw = adata`. I then extract 3 separate `adata`s like `adata_1 = adata[adata.obs['type'] == 1]`. `adata_1.raw` is preserved etc. Then I do some batch correction on each of the 3 `adata`s separately, then I concatenate to create the full dataset again via `Anndata.concatenate(adata_1, adata_2, adata_3, join="outer")` but I want to also preserve the `.raw` expression values in the sparse matrix.

**Comment by flying-sheep** [comment]
Hmm, in #94, @VolkerBergen extended it to layers.

Btw: custom attributes like your `name` are of course discarded! Nothing supports those, `pd.DataFrame` just has them as shortcut for column access.

**Comment by VolkerBergen** [comment]
Agree with @chris-rands; guess `.raw` should be concatenated just like the other attributes, which should be straightforward to implement like we did for `.layers`.

**Comment by chris-rands** [comment]
Thanks both, very naively, I'm testing this workaround, does it make sense?

```
adata_concat = Anndata.concatenate(*adatas, join="outer")
raws = [adata.raw.to_adata() for adata in adatas]
adata_concat.raw = Anndata.concatenate(*raws, join="outer")
```

**Comment by flying-sheep** [comment]
Yes. To turn this into a PR, you basically just have to

1. Move lines 2-3 into `concatenate`
2. Pass the parameters on to the inner `concatenate` call
3. Some adatas might have `.raw`s, you need to deal with that
4 There needs to be a way to check which raw came from which adata, especially because .raws don’t have `obs`.
5. Add some tests that check if points 2-4 work as intended

**Comment by chris-rands** [comment]
Thanks, sounds fairly straightforward, I'm happy to work on a PR for this

**Comment by LuckyMD** [comment]
so if you have different gene sets in the `.raw` objects... what do you do?

**Comment by VolkerBergen** [comment]
> so if you have different gene sets in the `.raw` objects... what do you do?
Merge to shared `var_names` and throw a warning?

**Comment by LuckyMD** [comment]
Suggestion here seems to be an outer join, which would pad with zeros if i understand correctly. That sounds dangerous. I’d go for @VolkerBergen’s suggestion as well if we want to do this.

**Comment by chris-rands** [comment]
I had a go at this (see the PR); hoping that since I'm using the existing `concatenate` function, it would appropriately handle these things, but I might be wrong

**Comment by chris-rands** [comment]
Closed via #265

**Comment by watiss** [comment]
I am facing a similar issue now. I am concatenating two identical anndatas (as a test). They have the same var dataframe under `adata.raw.var` and `adata2.raw.var`. However, `adata3`, which is the results of `adata3 = adata.concatenate(adata2, join="outer")` has a dataframe in `adata3.raw.var` that is missing the columns (only the index is there). Is this expected?
